staging/src/k8s.io/apiserver/pkg/storage/cacher/util.go

```diff
@@ -44,3 +48,41 @@ func hasPathPrefix(s, pathPrefix string) bool {
 	}
 	return false
 }
+
+// storageResourceVersionGuard ensures that
+// storage.GetCurrentResourceVersionFromStorage is computed
+// only once.
+//
+// note that this struct is not thread safe.
+type storageResourceVersionGuard struct {
+	computed                bool
+	computedResourceVersion uint64
+	computedErr             error
+
+	ctx               context.Context
+	underlyingStorage storage.Interface
+	newListFunc       func() runtime.Object
+	resourcePrefix    string
+	objectType        string
+}
+
+// newStorageResourceVersionGuard creates a new instance of the storage guard.
+// This function should be called for each new request.
+func newStorageResourceVersionGuard(ctx context.Context, underlyingStorage storage.Interface, newListFunc func() runtime.Object, resourcePrefix, objectType string) *storageResourceVersionGuard {
+	return &storageResourceVersionGuard{
+		ctx:               ctx,
```

Comment:
context isn't supposed to be stored, the `getCurrentResourceVersionFromStorageOnce` could take a context instead

Reply:
`newStorageResourceVersionGuard` will be created per request, so the `ctx` will be per request but I don't mind changing the signature of the method to accept the `ctx`.